### PR TITLE
Changes to use pathlib

### LINKED ISSUE #1991: The tests leave find_spot_server.py processes running
After running the DIALS tests to completion, I see two `find_spots_server.py` processes still running:
```
$ ps aux | grep find_spots_server.py | grep -v grep
fcx32934 27277  0.0  0.3 470356 218528 pts/0   S    13:22   0:00 /home/fcx32934/sw/cctbx/build/../conda_base/bin/python3.8 /home/fcx32934/sw/cctbx/build/../modules/dials/command_line/find_spots_server.py port=50879 nproc=3
fcx32934 27278  0.0  0.1 371176 108532 pts/0   S    13:22   0:00 /home/fcx32934/sw/cctbx/build/../conda_base/bin/python3.8 /home/fcx32934/sw/cctbx/build/../modules/dials/command_line/find_spots_server.py port=50879 nproc=3
```

### LINKED ISSUE #2037: Remove deprecated method in test
```
DeprecationWarning: The DataFetcher currently returns py.path.local() objects. This will in the future change to pathlib.Path() objects. You can either add a pathlib=True argument to obtain a pathlib.Path() object, or pathlib=False to silence this warning for now.
```

One or two of these (OK, hundreds) were emitted when I ran the tests. Seems like it wants us to do a thing. Please could we do a thing? Seems very busy work / find and replace code.

### PR DESCRIPTION
**Pull request by graeme-winter**
This is an attempt to silence some deprecation warnings

For #2037

[Edit @Anthchirp: Fixes #2037, Fixes #1991]

### CODE REVIEW

**Comment by graeme-winter** [comment]
148 instances of this so would like to get this right once before changing it everywhere.

File: tests/util/test_slice.py [hunk 1]
```diff
@@ -21,8 +21,8 @@ def test_slice_experiments():
 
 
 def test_slice_experiments_centroid_test_data(dials_data):
-    files = dials_data("centroid_test_data").listdir("*.cbf", sort=True)
-    experiments = ExperimentListFactory.from_filenames(f.strpath for f in files)
+    files = sorted(dials_data("centroid_test_data", pathlib=True).glob("*.cbf"))
+    experiments = ExperimentListFactory.from_filenames(str(f) for f in files)
```
Comment (rjgildea) [inline tests/util/test_slice.py:25]:
```suggestion
    experiments = ExperimentListFactory.from_filenames(files)
```

Comment (rjgildea) [reply tests/util/test_slice.py:25]:
```
>>> ExperimentListFactory.from_filenames([pathlib.Path("/dls/science/groups/scisoft/DIALS/dials_data/x4wide/X4_wide_M1S4_2_0001.cbf")])
ExperimentList([<dxtbx_model_ext.Experiment object at 0x7febd482b280>])
```

**Comment by benjaminhwilliams** [comment]
I'm not sure if the `run_in_tmpdir` → `run_in_tmp_path` changes are entirely correct.  They're fairly numerous so I'll leave it to the CI checks to tell us.

There is some code duplication with the changes in cctbx/dxtbx#505.  It might be better to replace the fixture definition in `dials.conftest` with
```from dxtbx.conftest import run_in_tmp_path  # noqa; lgtm; exported symbol```
and do the same in `xia2.conftest`.

**Comment by Anthchirp** [comment]
All boxes in #2037 are now ticked

**Pull request merged**